entrance_rando.py

```diff
@@ -0,0 +1,429 @@
+import itertools
+import logging
+import random
+import time
+from collections import deque
+from typing import Callable, Dict, Iterable, List, Tuple, Set, Optional
+
+from BaseClasses import CollectionState, Entrance, Region, EntranceType
+from Options import Accessibility
+from worlds.AutoWorld import World
+
+
+class EntranceRandomizationError(RuntimeError):
+    pass
+
+
+class EntranceLookup:
+    class GroupLookup:
+        _lookup: Dict[int, List[Entrance]]
+
+        def __init__(self):
+            self._lookup = {}
+
+        def __len__(self):
+            return sum(map(len, self._lookup.values()))
+
+        def __bool__(self):
+            return bool(self._lookup)
+
+        def __getitem__(self, item: int) -> List[Entrance]:
+            return self._lookup.get(item, [])
+
+        def __iter__(self):
+            return itertools.chain.from_iterable(self._lookup.values())
+
+        def __str__(self):
+            return str(self._lookup)
+
+        def __repr__(self):
+            return self.__str__()
```

Comment:
```suggestion
        def __repr__(self):
            return str(self._lookup)
```
`__str__` automatically defaults to `__repr__`
If you just define `__repr__` then `__str__` will automatically be the same thing.